Join-Dependent Predicate Duplication

ggevay opened the pull request:
There are cases where a traditional predicate pushdown is not possible, but it's still possible to create an extra filter on some join inputs, and thereby reduce the data volume that goes into the join.

For example:
```
SELECT * FROM t1, t2 WHERE t1.f1 = t2.f1 AND ((t1.f2 = 3 AND t2.f2 = 4) OR (t1.f2 = 5 AND t2.f2 = 6));
```
Here, we can pre-filter `t1` with `t1.f2 = 3 OR t1.f2 = 5`, and `t2` with `t2.f2 = 4 OR t2.f2 = 6`.

Similar situations occur in TPC-H Q07, Q19, and chbench Q07, Q19.

### Motivation

  * This PR adds a known-desirable feature: #14501, and #7409 (partially).

### Tips for reviewer

In TPC-H Q19 and chbench Q19 some of the newly created filters get lifted to after the join by `join_implementation.rs`, which doesn't looks so good at first glance. However:
- I presume that the rendering might still be able to push those to the join closure that is just after that input, which means that we'll filter with them before joining in other inputs.
- Even if they end up in the final closure, evaluating the expression might still be slightly faster, because the lifted filters (fortunately) come before the original predicate, and they are simpler than the original predicate.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Performance improvement, see above.

antiguru commented on src/expr/src/relation/join_input_mapper.rs:
```diff
@@ -354,6 +354,93 @@ impl JoinInputMapper {
         None
     }
+
+    /// Try to find a consequence `c` of the given expression `e` for the given input.
+    ///
+    /// If we return `Some(c)`, that means
+    ///   1. `c` uses only columns from the given input;
+    ///   2. if `c` doesn't hold on a row of the input then `e` also wouldn't hold;
+    ///   3. if `c` holds on a row of the input then `e` might or might not hold.
+    /// 1. and 2. means that if we have a join with predicate `e` then we can use `c` for
+    /// pre-filtering a join input before the join. However, 3. means that `e` shouldn't be deleted
+    /// from the join predicates, i.e., we can't do a "traditional" predicate pushdown.
+    ///
+    /// Note that "`c` is a consequence of `e`" is the same thing as 2., see
+    /// <https://en.wikipedia.org/wiki/Contraposition>
+    ///
+    /// Example: For
+    /// `(t1.f2 = 3 AND t2.f2 = 4) OR (t1.f2 = 5 AND t2.f2 = 6)`
+    /// we find
+    /// `t1.f2 = 3 OR t1.f2 = 5` for t1, and
+    /// `t2.f2 = 4 OR t2.f2 = 6` for t2.
+    ///
+    /// Further examples are TPC-H Q07, Q19, and chbench Q07, Q19.
```
Please explain the parameters. Thanks for adding such a detailed description of what the function does!

ggevay replied:
Done

antiguru commented on src/expr/src/relation/join_input_mapper.rs:
```diff
@@ -354,6 +354,93 @@ impl JoinInputMapper {
+        if let Some(localized) =
+            self.try_map_to_input_with_bound_expr(expr.clone(), index, equivalences)
+        {
+            return Some(localized);
```
No need for `return` here and below, all branches return a value.

ggevay replied:
Done

antiguru commented on src/expr/src/relation/join_input_mapper.rs:
```diff
@@ -354,6 +354,93 @@ impl JoinInputMapper {
+                            })
+                        })
+                        .collect_vec();
+                    if consequences_per_arg.iter().any(|c| matches!(c, None)) {
```
```suggestion
                    if consequences_per_arg.iter().any(Option::is_none) {
```

ggevay replied:
Fixed

antiguru commented on src/expr/src/relation/join_input_mapper.rs:
```diff
@@ -354,6 +354,93 @@ impl JoinInputMapper {
+                            })
+                        })
+                        .collect_vec();
+                    if consequences_per_arg.iter().all(|c| matches!(c, None)) {
```
```suggestion
                    if consequences_per_arg.iter().all(Option::is_none) {
```

ggevay replied:
Fixed

antiguru commented on src/expr/src/relation/join_input_mapper.rs:
```diff
@@ -354,6 +354,93 @@ impl JoinInputMapper {
+                                self.consequence_for_input(or_arg, index, equivalences)
+                            })
+                        })
+                        .collect_vec();
+                    if consequences_per_arg.iter().any(|c| matches!(c, None)) {
+                        return None;
+                    } else {
+                        return Some(MirScalarExpr::CallVariadic {
+                            func: VariadicFunc::Or,
+                            exprs: consequences_per_arg
+                                .into_iter()
+                                .map(|c| c.unwrap())
+                                .collect_vec(),
+                        });
+                    }
```
```suggestion
                        .collect::<Option<Vec<_>>>()?;
                    Some(MirScalarExpr::CallVariadic {
                        func: VariadicFunc::And,
                        exprs: consequences_per_arg,
                    })
```

antiguru replied:
Can do the same trick above! `collect::<Option<Vec<_>>>()` will return `None` if any element is `None` and otherwise `Some(Vec))`. This also saves an intermediate allocation.

ggevay replied:
That's a nice trick, but I'm not sure whether it would make this function more readable. The problem is that we can do this trick only in the `Or` branch of the top-level match, but not the `And` branch (because that uses `.all` instead of `.any`). If we do it only in the `Or` branch, that would make the `Or` and `And` branches differ more, which would obscure the correspondence between them.

antiguru replied:
Oh I see. I think for the `all` case, you could filter and collect and check is the result is non-empty. I think avoiding intermediate allocations is important, but I don't want to be prescriptive!

ggevay replied:
Ok, I changed both branches.

antiguru commented on src/transform/src/predicate_pushdown.rs:
```diff
@@ -281,6 +281,17 @@ impl PredicatePushdown {
+                                        .consequence_for_input(
+                                            predicate.clone(),
+                                            index,
+                                            &equivalences[..],
```
Non need for `[..]`, it'll dereference to `&[Vec<_>]` without the range expression.

ggevay replied:
Fixed

antiguru reviewed (commented):
This seems to look good! I can't really comment on whether it does the right thing (it looks like but I'd love to have someone with more optimizer experience look at it.) I left some stylistic comments.

wangandi commented on test/sqllogictest/transform/predicate_pushdown.slt:
```diff
@@ -311,3 +311,93 @@ Query:
 
 EOF
+
+# Join-Dependent Predicate Duplication (JoinInputMapper::consequence_for_input)
+
+query T multiline
+EXPLAIN
+SELECT * FROM t1, t2 WHERE t1.f1 = t2.f1 AND ((t1.f2 = 3 AND t2.f2 = 4) OR (t1.f2 = 5 AND t2.f2 = 6));
```
You might as well use the new explain interface to print out these plans to save effort when switching between old and new explain.

ggevay replied:
I guess the entire `predicate_pushdown.slt` will be rewritten using `--rewrite-results`, so the new tests won't really add extra effort.

wangandi commented:
> I presume that the rendering might still be able to push those to the join closure that is just after that input, which means that we'll filter with them before joining in other inputs.

You don't need to presume. You can create `EXPLAIN PHYSICAL PLAN AS TEXT` or `EXPLAIN PHYSICAL PLAN AS JSON` tests to see if the filters are in the join closure that is just after that input.

wangandi commented on src/expr/src/relation/join_input_mapper.rs:
```diff
@@ -354,6 +355,90 @@ impl JoinInputMapper {
+        index: usize,
+        equivalences: &[Vec<MirScalarExpr>],
+    ) -> Option<MirScalarExpr> {
+        if let Some(localized) =
+            self.try_map_to_input_with_bound_expr(expr.clone(), index, equivalences)
+        {
+            Some(localized)
+        } else {
```
To avoid calling `try_map_to_input_with_bound_expr` and `expr.clone()` multiple times, we can split `try_map_to_input_with_bound_expr` into two functions: 1) remaps the expression to refer to the given input whenever possible 2) checks that the expression has been entirely remapped to a given input. Then this method only needs to call the expression remapping method once, and the recursion only needs to check that the subexpression of the result all belongs to the given input.

ggevay replied:
Good idea, thx!

aalexandrov replied:
The `expr.clone()` was the only thing that bothered me at first glance. Great to see that there is a workaround for that!

wangandi commented on src/expr/src/relation/join_input_mapper.rs:
```diff
@@ -354,6 +355,90 @@ impl JoinInputMapper {
+                    func: VariadicFunc::Or,
+                    exprs: or_args,
+                } => {
+                    // Each OR arg should provide a consequence. It they do, we OR them.
```
```suggestion
                    // Each OR arg should provide a consequence. If they do, we OR them.
```

ggevay replied:
Fixed

philip-stoev commented:
@ggevay do you mind if I test this tomorrow?

ggevay commented:
@philip-stoev That would be great, thx!

ggevay commented:
Thanks, @wangandi! I addressed the comments.

>> I presume that the rendering might still be able to push those to the join closure that is just after that input, which means that we'll filter with them before joining in other inputs.

> You don't need to presume. You can create EXPLAIN PHYSICAL PLAN AS TEXT or EXPLAIN PHYSICAL PLAN AS JSON
tests to see if the filters are in the join closure that is just after that input.

Hm, good idea! I added a test with the physical plan of a delta join.

philip-stoev commented:
Item No 1. This query has a `Distinct` and a `Union` that are missing in the prior commit.

```
materialize=> EXPLAIN  SELECT * FROM t2 AS a1 left JOIN t2 AS a2 USING ( f1 ) WHERE a1 . f1 = 1 AND a2 . f1 = 2  OR a1 .f2 = 3 AND a1 .f2 = 4;
%0 = Let l0 =
| Get materialize.public.t2 (u4)
| ArrangeBy (#0)

%1 = Let l1 =
| Join %0 %0 (= #0 #2)
| | implementation = DeltaQuery
| |   delta %0 %0.(#0)
| |   delta %0 %0.(#0)
| Filter (#0) IS NOT NULL
| Project (#0, #1, #3)

%2 =
| Get %1 (l1)
| Filter (((#0 = 2) AND (#0 = 1)) OR ((#1 = 3) AND (#1 = 4)))

%3 =
| Get %1 (l1)
| Map (#0 = 1), ((#1 = 3) AND (#1 = 4))
| Filter (#3 OR #4), (#4 OR (#3 AND null))
| Project (#0, #1)
| Distinct group=(#0, #1)
| Negate

%4 =
| Get materialize.public.t2 (u4)
| Map (#0 = 1), ((#1 = 3) AND (#1 = 4))
| Filter (#2 OR #3), (#3 OR (#2 AND null))
| Project (#0, #1)
| Distinct group=(#0, #1) <- THIS HERE

%5 =
| Union %3 %4 <- AND THIS HERE

%6 =
| Get materialize.public.t2 (u4)
| ArrangeBy (#1, #0)

%7 =
| Join %5 %6 (= #0 #2) (= #1 #3)
| | implementation = Differential %5 %6.(#1, #0)
| Map (#0 = 1), ((#1 = 3) AND (#1 = 4)), null
| Filter (#4 OR #5), (#5 OR (#4 AND null))
| Project (#0, #1, #6)

%8 =
| Union %2 %7
```

Dataset in https://github.com/MaterializeInc/materialize/issues/14905

philip-stoev commented:
Item No 2. This query has a `Map` operator in this branch:

```
materialize=> EXPLAIN  SELECT * FROM t2 a1  JOIN t1  USING ( f1  )  WHERE (f1  IN (  1, 2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 , 10  )  AND a1 .f2  IN (  6  ,  4  ))  OR f1  IN (  0    )   ;
%0 =
| Get materialize.public.t2 (u4)
| ArrangeBy (#0)

%1 =
| Get materialize.public.t1 (u1)
| ArrangeBy (#0)

%2 =
| Join %0 %1 (= #0 #2)
| | implementation = DeltaQuery
| |   delta %0 %1.(#0)
| |   delta %1 %0.(#0)
| Map (#0 = 0), (#0 = 2), (#0 = 3), (#0 = 4), (#0 = 5), (#0 = 6), (#0 = 7), (#0 = 8), (#0 = 9), (#0 = 10), (#0 = 1)
| Filter (#0) IS NOT NULL, (#4 OR #5 OR #6 OR #7 OR #8 OR #9 OR #10 OR #11 OR #12 OR #13 OR #14), (#4 OR ((#5 OR #6 OR #7 OR #8 OR #9 OR #10 OR #11 OR #12 OR #13 OR #14) AND ((#1 = 4) OR (#1 = 6))))There is no short-cut tha
| Project (#0, #1, #3)
```

So I was wondering:
- why are we not joining the "join-with-constant-source" trick that kicks in in other cases
- With the `Map` *all* the comparisons between f1 and the individual items in the IN list are performed, right? In this case, there is no short-cut that will avoid computing the rest of the comparisons if a given one evaluates to True.

philip-stoev reviewed (approved):
I ran the RQG with two grammars -- one that creates random nestings of ANDs and ORs and IN lists and a TPC-*-like grammar that creates queries over the usual orders/lineitem/etc tables.

There were no wrong results in either case. I examined a zillion query plans that have changed and all the changes fall into the "Items" comments in the PR.

ggevay commented:
Thank you very much for the thorough testing @philip-stoev!

Item No 1.: This is actually an issue not in this PR, but in the literal constraints detection, so I will open a separate PR to fix it.

Item No 2.: At the moment, we are doing the "join-with-constant-source" trick only when the filter (or actually the MFP) is right after a `Get`. Predicate pushdown makes this to be the case:
```
Project (#0, #1, #3)
  Join on=(#0 = #2)
    Filter (#0) IS NOT NULL AND ((#1 = 4) OR (#1 = 6)) AND ((#0 = 2) OR (#0 = 3) OR (#0 = 4) OR (#0 = 5) OR (#0 = 6) OR (#0 = 7) OR (#0 = 8) OR (#0 = 9) OR (#0 = 10) OR (#0 = 1))
      Get materialize.public.t2
    Filter (#0) IS NOT NULL AND ((#0 = 2) OR (#0 = 3) OR (#0 = 4) OR (#0 = 5) OR (#0 = 6) OR (#0 = 7) OR (#0 = 8) OR (#0 = 9) OR (#0 = 10) OR (#0 = 1))
      Get materialize.public.t1
```
But then `join_implementation` lifts the filters, because it wants to use the existing indexes for a delta join, and if there are extra filters there, then it cannot use the existing index. So after `join_implementation`, the filter is no longer right after the `Get`.
```
Project (#0, #1, #3)
  Project (#0..=#3)
    Filter (#0) IS NOT NULL AND ((#0 = 2) OR (#0 = 3) OR (#0 = 4) OR (#0 = 5) OR (#0 = 6) OR (#0 = 7) OR (#0 = 8) OR (#0 = 9) OR (#0 = 10) OR (#0 = 1)) AND ((#1 = 4) OR (#1 = 6)) AND (#2) IS NOT NULL AND ((#2 = 2) OR (#2 = 3) OR (#2 = 4) OR (#2 = 5) OR (#2 = 6) OR (#2 = 7) OR (#2 = 8) OR (#2 = 9) OR (#2 = 10) OR (#2 = 1))
      Map ()
        Join on=(#0 = #2) type=delta
          ArrangeBy keys=[[#0]]
            Get materialize.public.t2
          ArrangeBy keys=[[#0]]
            Get materialize.public.t1
```
I wrote a comment now about this here:
https://github.com/MaterializeInc/materialize/issues/14500#issuecomment-1257179631
